### ISSUE TITLE
feaLib: sorting of LigatureCaretByIndex rules?

### ISSUE THREAD

**Issue by cmyr** [fonttools/fonttools#3608]
When generating a ligature caret table, feaLib sorts the `CaretValue`s. This definitely makes sense for values that are explicit coordinates, since we expect the smallest value to represent the first caret. In the case of point indices, though, it's not obvious that this is correct, since in an outline an increase in point index does not necessarily represent an increase in the coordinate value for that index.

My instinct would be to sort the values when they are coordinates, but to leave the user's ordering unchanged if they are indices. Am I missing anything?

**Comment by anthrotype** [comment]
are you referring to the `sorted(points)` in this, right?

https://github.com/fonttools/fonttools/blob/16dbe3f5e22555c88a63f85e753c1c13311ded2b/Lib/fontTools/otlLib/builder.py#L2576-L2589

> My instinct would be to sort the values when they are coordinates, but to leave the user's ordering unchanged if they are indices

I think you're right, it looks like an oversight

**Comment by cmyr** [comment]
yes, exactly. A very reasonable oversight, i'd be surprised if this format is used in any font we compile..

**Comment by anthrotype** [comment]
ufo2ft only generates LigatureCaretByPosStatement, not the ByIndex sort

**Comment by behdad** [comment]
Does the sorting consider right-to-left ligatures?

**Comment by anthrotype** [comment]
> Does the sorting consider right-to-left ligatures?

good question. I think it doesn't /cc @simoncozens ?

**Comment by anthrotype** [comment]
the ufo2ft gdefFeatureWriter also sorts them:

https://github.com/googlefonts/ufo2ft/blob/bb79cae53f1c160c7174ebef0d463c7a28a7552a/Lib/ufo2ft/featureWriters/gdefFeatureWriter.py#L79

**Comment by behdad** [comment]
Would *not* sorting has any downside? I expect the designer to put the right thing in there to begin with.

**Comment by behdad** [comment]
cc @khaledhosny

**Comment by anthrotype** [comment]
> I expect the designer to put the right thing in there to begin with

yeah. Or the gdefFeatureWriter could make use of the indices in the caret_1, caret_2 etc. anchor names and sort by those. I think right now it completely ignores them (e.g. `caret_` without any suffix works)

**Comment by anthrotype** [comment]
oh maybe they have to be sorted?

https://learn.microsoft.com/en-us/typography/opentype/spec/gdef#ligature-caret-list-table

> 
Offset16 | caretValueOffsets[caretCount] | Array of offsets to caret value tables, from beginning of LigGlyph table, **in increasing coordinate order**.
-- | -- | --

**Comment by behdad** [comment]
> oh maybe they have to be sorted?

@khaledhosny what's your implementation of this in Pango / LibreOffice, and what does CoreText do? Do they expect the carets in a RTL ligature (say, Lam-Lam-Heh) to be right-to-left, or left-to-right?

**Comment by khaledhosny** [comment]
The three implementations work fine with ligature caret positions for RTL ligatures generated by ufo2ft/feaLib, so whatever they are doing now seems to be good.
